plugins/crd/crd_queue.go

```diff
@@ -0,0 +1,270 @@
+// Copyright (c) 2018 Cisco and/or its affiliates.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at:
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// //go:generate protoc -I ./model/pod --go_out=plugins=grpc:./model/pod ./model/pod/pod.proto
+
+package netmeshplugincrd
+
+import (
+	"fmt"
+
+	"k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/client-go/tools/cache"
+)
+
+// This file contains the work queue backends for each of the CRDs we create in the
+// plugin AfterInit() call.
+
+func networkserviceWork(plugin *Plugin) {
+	for {
+		// We read a message off the queue ...
+		key, shutdown := queueNS.Get()
+
+		// If the queue has been shut down, we should exit the work queue here
+		if shutdown {
+			plugin.informerStopCh <- struct{}{}
+			return
+		}
+
+		// Convert the queue item into a string. If it's not a string, we'll
+		// simply discard it as invalid data and log a message.
+		var strKey string
+		var ok bool
+		if strKey, ok = key.(string); !ok {
+			runtime.HandleError(fmt.Errorf("key in queue should be of type string but got %T. discarding", key))
+			return
+		}
+
+		// We define a function here to process a queue item, so that we can
+		// use 'defer' to make sure the message is marked as Done on the queue
+		func(key string) {
+			var obj interface{}
+			var err error
+
+			defer queueNS.Done(key)
+
+			// Attempt to split the 'key' into namespace and object name
+			namespace, name, err := cache.SplitMetaNamespaceKey(strKey)
+
+			if err != nil {
+				plugin.Log.Errorf("Error splitting meta namespace key into parts: %s", err.Error())
+				runtime.HandleError(fmt.Errorf("Error splitting meta namespace key into parts: %s", err.Error()))
+				// This is a soft-error, we merely want to forget it and mark it done on the queue
+				queueNS.Forget(key)
+				return
+			}
+
+			plugin.Log.Infof("Read item '%s/%s' off workqueue. Processing...", namespace, name)
+
+			// Retrieve the latest version in the cache of this alert
+			obj, err = plugin.sharedFactoryNS.Networkservice().V1().NetworkServices().Lister().NetworkServices(namespace).Get(name)
+
+			if err != nil {
+				plugin.Log.Errorf("Error getting object '%s/%s' from api: %s", namespace, name, err.Error())
+				runtime.HandleError(fmt.Errorf("Error getting object '%s/%s' from api: %s", namespace, name, err.Error()))
+				// This is a hard-error, we'll raise the flag so the plugin catches this and shuts down
+				queueNS.Forget(key)
+				plugin.queueError <- true
+				return
```

Comment:
For some reason, on a CRD delete, this is triggered, meaning the factory can't find the object. An example of the error I get is as follows:

```
time="2018-05-30 18:51:03.41696" level=info msg="Read item 'default/yogi-bear' off workqueue. Processing..." loc="crd/crd_queue.go(67)" logger=netmeshcrd
time="2018-05-30 18:51:03.41759" level=info msg="Dequeuing interface of type networkservices" loc="crd/crd_queue.go(70)" logger=netmeshcrd
ERROR: logging before flag.Parse: E0530 18:51:03.417636       1 crd_queue.go:74] error getting object 'default/yogi-bear' from api: networkservice.networkservicemesh.io "yogi-bear" not found
```

We can merge this as-is and fix this after, or someone can help me eyeball what might be causing this.